### ISSUE TITLE
Problems with testing Deferred deep linking

### ISSUE THREAD

**Issue by avinz** [adjust/android_sdk#237]
Hello,
I am trying to test Deferred deep linking in application, so after install user will pass directly to desired page. 

AdjustReferrerReceiver is added to manifest.

So we have a link :  https://app.adjust.com/blabla1?deep_link=some://foo/bar

(if this link is used when app is installed, deeplinking works as expected)

The way I'm trying to test it now : 
1)Remove the app from device
2)Forget device - https://app.adjust.com/forget_device?app_token={token}&gps_adid={gpsadid}
3)Install application but don't launch
4)send command via adb shell : 
`am broadcast -a com.android.vending.INSTALL_REFERRER -n packet.packet/com.adjust.sdk.AdjustReferrerReceiver --es "referrer" "adjust_reftag%3Dc6wfjZrfVrecVk%26utm_source%3DSome"`
In LogCat i get following log : 
>11-04 15:37:29.316 21853-23336 D/Adjust: Attribution file not found
11-04 15:37:29.333 21853-23336 D/Adjust: Activity state file not found
11-04 15:37:29.336 21853-23336 D/Adjust: Session Callback parameters file not found
11-04 15:37:29.337 21853-23336 D/Adjust: Session Partner parameters file not found
11-04 15:37:29.778 21853-23336 I/Adjust: Google Play Services Advertising ID read correctly at start time
11-04 15:37:29.781 21853-23336 V/Adjust: Foreground timer configured to fire after 60.0 seconds of starting and cycles every 60.0 seconds
11-04 15:37:29.794 21853-23356 D/Adjust: Package queue file not found
11-04 15:37:29.820 21853-23336 V/Adjust: Reading query string (adjust_reftag={reftag}&utm_source=Some)
11-04 15:37:29.840 21853-23358 D/Adjust: Added sdk_click 1
11-04 15:37:29.840 21853-23358 V/Adjust: Path:      /sdk_click
                                         ClientSdk: android4.10.2
                                         Parameters:
                                         	app_token        {token}
                                         	attribution_deeplink 1
                                         	click_time       2016-11-04T15:37:29.317Z+0200
                                         	created_at       2016-11-04T15:37:29.820Z+0200
                                         	environment      sandbox
                                         	event_buffering_enabled 0
                                         	gps_adid         {gps_adid}
                                         	needs_response_details 0
                                         	referrer         adjust_reftag={reftag}&utm_source=Some
                                         	reftag           {reftag}
                                         	source           reftag
>                                         	tracking_enabled 1

After I launch the app this is the first response : 
>11-04 15:41:15.805 21853-27089 V/Adjust: Response: {"app_token":"{token}","adid":"{adid}","timestamp":"2016->11-04T13:41:18.602Z+0000","message":"Click tracked"}
>I/Adjust: Click tracked

Afterwards Install is tracked and thats it, no deep link further. 
May be I'm doing smth wrong, but I researched for a long time... Need help ^_^

**Comment by uerceg** [comment]
Hi @avinz 

Do you have attribution callback implemented in your app? If not, add at least empty implementation of it and retry the test flow. Deferred deep link should arrive then.

4.10.3 will contain a change where you won't need to have this callback implemented for deferred deep link to arrive.

Lemme know if it worked for you.

Cheers

**Comment by avinz** [comment]
Hello @uerceg 

Thank you very much for reply, now it started working :) 
But as I see (and seem written in docks) that there is no guarantee that deeplink event will arrive immediately to the application, sometime it arrives immediately, sometimes in one minute and sometimes only after I close and open app again. Is there a way to receive this event on app launch for sure?

**Comment by uerceg** [comment]
Hi @avinz

Good news. :)

As of SDK v4.10.0, deferred deep link is being delivered in the answer to the attribution answer coming from the backend. Flow looks pretty much like this: 

If your user have clicked on URL which contained `deep_link` info in it, he/she was redirected to the store, have downloaded and launched the app. In that moment, SDK sends initial session package to the backend and tracks an install. After we get the response from the backend that install has been tracked, if attribution callback is implemented in the app (this is the part which in next update won't be prerequisite for what goes next), SDK will ask the backend about your user's attribution info. If our backend before got the info that click, which your install is going to be matched with, contained any deep link info (or if we maybe got info from some other partners (for example Facebook) that deep link info for this install was found by them), this info will be sent in the response to the SDK and SDK will trigger deferred deep link callback if you have implemented it in the app. In case of Android, if you said `return true` in deferred deep link callback, we will also launch intent which will start the activity which handles your deep link scheme and info about the link will be available in that activity as well as part of either `onCreate` or `onNewIntent` lifecycle methods, depending on what's that activity `android:launchMode` setting.

This is how it works and in theory it should always work like that.

**Comment by avinz** [comment]
@uerceg 

After app is launched I receive +- 5 same events : 

>11-05 13:43:39.342 14181-16788 V/Adjust: Attribution timer fired
11-05 13:43:39.342 14181-16788 V/Adjust: Path:      attribution
                                         ClientSdk: android4.10.2
                                         Parameters:
                                         	app_token        {token}
                                         	attribution_deeplink 1
                                         	created_at       2016-11-05T13:42:51.313Z+0200
                                         	environment      sandbox
                                         	event_buffering_enabled 0
                                         	gps_adid         {adid}
                                         	needs_response_details 1
                                         	tracking_enabled 1
11-05 13:43:39.694 14181-16788 V/Adjust: Response: {"app_token":"{token}","adid":"{adid}","timestamp":"2016-11-05T11:43:43.308Z+0000","message":"Attribution found","attribution": {"tracker_token":"{ttoken}","tracker_name":"Organic","network":"Organic"},"ask_in":200}
11-05 13:43:39.695 14181-16788 D/Adjust: Wrote Activity state: ec:0 sc:1 ssc:1 sl:0.0 ts:0.0 la:11:12:13 uuid:{uuid}
11-05 13:43:39.695 14181-16788 D/Adjust: Waiting to query attribution in 0.2 seconds
11-05 13:43:39.695 14181-16788 V/Adjust: Attribution timer canceled
>11-05 13:43:39.695 14181-16788 V/Adjust: Attribution timer starting. Launching in 0.2 seconds


Only after I closed and opened app again the same request was made but with answer deeplink also arrived. 

>11-05 13:44:45.480 14181-18698 V/Adjust: Response: {"app_token":"{token}","adid":"{adid}","timestamp":"2016-11-05T11:44:49.106Z+0000","message":"Attribution found","attribution":{"tracker_token":"{ttoken}","tracker_name":"Vacacy185","network":"Vacacy185","deeplink":"app://foo/bar?adjust_no_sdkclick=1"}}
11-05 13:44:45.480 14181-18698 I/Adjust: Attribution found
11-05 13:44:45.486 14181-18698 D/Adjust: Wrote Activity state: ec:4 sc:1 ssc:2 sl:70.2 ts:68.0 la:11:12:13 uuid:{udid}
11-05 13:44:45.489 14181-16731 D/Adjust: Wrote Attribution: tt:{ttoekn} tn:Vacacy185 net:Vacacy185 cam:null adg:null cre:null cl:null
11-05 13:44:45.495 14181-16731 I/Adjust: Deferred deeplink received (app://foo/bar?adjust_no_sdkclick=1)
11-05 13:44:45.858 14181-14181 D/App: Attribution changed
11-05 13:44:45.858 14181-14181 D/App: Launching deeplink
11-05 13:44:45.858 14181-14181 I/Adjust: Open deferred deep link (app://foo/bar?>adjust_no_sdkclick=1)

Cannot figure out what is the logic when adjust fires deeplink back... Though request with deeplink was made before the app was launched.

**Comment by avinz** [comment]
@uerceg So this issue is still present and adjust doesn't guarantee do send deeplink just after app install?

**Comment by uerceg** [comment]
Hi @avinz 

And sorry for the delayed answer. So, are you facing this problem with Android 4.10.4?

I would really like to see how log output looks like for your test device if you do the following scenario:

1. Uninstall the app completely from your test device.
2. Forget your test device from our backend by hitting the forget URL.
3. Make a click to some URL which has deep link information in it.
4. Install the app to your device and launch it with SDK environment set to sandbox and log level set to verbose.

After that, please paste in here the logs you're seeing (feel free to hide app/event tokens, device IDs that you think as the ones which shouldn't be shared in here).

**Comment by avinz** [comment]
@uerceg Logs witch Ive posted in previous post are same as I receive now.

**Comment by uerceg** [comment]
@avinz Can you show me how does the tracker URL which contains deep link info looks like which you are using to test deferred deep linking scenario?

**Comment by avinz** [comment]
@uerceg If I understood you correct, this sis the link with I get from adjust link. 

https://play.google.com/store/apps/details?id=packet.packet&referrer=adjust_reftag%3Dreftag%26utm_source%3Dappname%253A%252F%252Ffirstpart%252Fsecondpart

I use reftag and source in adb shell request

am broadcast -a com.android.vending.INSTALL_REFERRER -n app.app/com.adjust.sdk.AdjustReferrerReceiver --es "referrer" "adjust_reftag%3Dreftag%26utm_source%3Dappname"

**Comment by uerceg** [comment]
Ah, no no. This is not how you are testing the deferred deep links. Broadcast receiver which gets install referrer info (and forwards them to our backend) is used to properly attribute the user which got redirected to the Play Store by clicking on some tracker URL.

In order to test deferred deep linking, first you need to enable it for your app, like described in here: https://github.com/adjust/android_sdk/blob/master/README.md#deeplinking

After you have enabled deep linking and assigned certain scheme name which your app will handle, you need to make tracker URL which will contain `deep_link` parameter. After you build it, it should look something like this (all other potential parameters stripped from it, just `deep_link` parameter in it:

https://app.adjust.com/abc123?deep_link=yourScheme://something_something_etc

More on this can be found in here: https://docs.adjust.com/en/deeplinking/#deferred-deeplinking

Anyway, value of `deep_link` parameter needs to be URL encoded, so if your idea is that URL should look something like this: https://app.adjust.com/abc123?deep_link=yourScheme://something, link which you need to click onto is this one: https://app.adjust.com/abc123?deep_link=yourScheme%3a%2f%2fsomething

So, you need to have this link on some web page which you will be able to open from your mobile web browser and make a click on it.

Then follow the flow I described in here: https://github.com/adjust/android_sdk/issues/237#issuecomment-264458215

And let me know if it worked for you.

Ping me if you need any further assistance with this.

**Comment by avinz** [comment]
@uerceg So it only possible to test by installing only from market after clicking on adjust link with deeplink in it?

**Comment by uerceg** [comment]
@avinz Not necessarily from the market. You don't even need to have app on the Play Store which supports this. You can do all locally with your debug version of app which has deep linking handling enabled, adjust SDK log level set to verbose and environment to sandbox. Once you click adjust link, if your app in the store, you will be redirected to the store on the page of your app, but you don't need to download that version. You can just close the store and install your current version of app which you are testing from Android Studio. The effect is the same, deferred deep link will be delivered to the app as well and you can check the whole flow in the logs.

**Comment by avinz** [comment]
@uerceg I've made following steps :

1)Forgot device with link "http://app.adjust.com/forget_device?app_token={app_token}&android_id={device_id}}" 
2)Clicked on link on the device https://app.adjust.com/{tracker_id}?deep_link={package}%3a%2f%2f{some}%2f{some} (this link works with insalled app)
3)Installed application from Studio and launched it. Mode - sandbox. Log - Verbose.

Here are the logs witch I received on launch.

>12-05 22:59:19.714 5005-8080 D/Adjust: Attribution file not found
12-05 22:59:19.722 5005-8080 D/Adjust: Activity state file not found
12-05 22:59:19.725 5005-8080 D/Adjust: Session Callback parameters file not found
12-05 22:59:19.725 5005-8080 D/Adjust: Session Partner parameters file not found
12-05 22:59:21.782 5005-8080 I/Adjust: Google Play Services Advertising ID read correctly at start time
12-05 22:59:21.825 5005-8080 V/Adjust: Foreground timer configured to fire after 60.0 seconds of starting and cycles every 60.0 seconds
12-05 22:59:21.898 5005-8208 D/Adjust: Package queue file not found
12-05 22:59:22.076 5005-8080 V/Adjust: Foreground timer starting
12-05 22:59:22.081 5005-8080 V/Adjust: Subsession start
12-05 22:59:22.197 5005-8229 D/Adjust: Added package 1 (session)
12-05 22:59:22.201 5005-8229 V/Adjust: Path:      /session
                                       ClientSdk: android4.10.4
                                       Parameters:
                                       	android_uuid     {uuid}
                                       	api_level        23
                                       	app_token        {app_token}
                                       	app_version      1.7.1
                                       	attribution_deeplink 1
                                       	country          US
                                       	cpu_type         armeabi-v7a
                                       	created_at       2016-12-05T22:59:22.090Z+0700
                                       	device_manufacturer LGE
                                       	device_name      Nexus 5
                                       	device_type      phone
                                       	display_height   1776
                                       	display_width    1080
                                       	environment      sandbox
                                       	event_buffering_enabled 0
                                       	fb_id            {fb_id}
                                       	gps_adid         {gps_aid}
                                       	hardware_name    M4B30X
                                       	language         en
                                       	needs_response_details 0
                                       	os_build         M4B30X
                                       	os_name          android
                                       	os_version       6.0.1
                                       	package_name     {package.package}
                                       	screen_density   high
                                       	screen_format    normal
                                       	screen_size      normal
                                       	session_count    1
                                       	tracking_enabled 1
                                       	vm_isa           arm
12-05 22:59:22.217 5005-8229 D/Adjust: Wrote Package queue: [session]
12-05 22:59:22.218 5005-8229 D/Adjust: Package handler wrote 1 packages
12-05 22:59:22.255 5005-8080 D/Adjust: Wrote Activity state: ec:0 sc:1 ssc:1 sl:0.0 ts:0.0 la:11:12:13 uuid:{uuid}
12-05 22:59:23.413 5005-8232 V/Adjust: Response: {"app_token":"{token}","adid":"{adid}","timestamp":"2016-12-05T15:59:23.747Z+0000","message":"Install tracked","attribution":{"tracker_token":"{tracker_token}","tracker_name":"{tracker_name}","network":"{tracker_name}"},"ask_in":2000}
12-05 22:59:23.414 5005-8232 I/Adjust: Install tracked
12-05 22:59:23.423 5005-8274 D/Adjust: Wrote Package queue: []
12-05 22:59:23.423 5005-8274 D/Adjust: Package handler wrote 0 packages
12-05 22:59:23.423 5005-8274 V/Adjust: Package handler can send
12-05 22:59:23.443 5005-8275 D/Adjust: Wrote Activity state: ec:0 sc:1 ssc:1 sl:0.0 ts:0.0 la:11:12:13 uuid:{uuid}
12-05 22:59:23.444 5005-8275 D/Adjust: Waiting to query attribution in 2.0 seconds
12-05 22:59:23.444 5005-8275 V/Adjust: Attribution timer canceled
12-05 22:59:23.444 5005-8275 V/Adjust: Attribution timer starting. Launching in 2.0 seconds
12-05 22:59:25.482 5005-8275 V/Adjust: Attribution timer fired
12-05 22:59:25.483 5005-8275 V/Adjust: Path:      attribution
                                       ClientSdk: android4.10.4
                                       Parameters:
                                       	app_token        {token}
                                       	attribution_deeplink 1
                                       	created_at       2016-12-05T22:59:21.862Z+0700
                                       	environment      sandbox
                                       	event_buffering_enabled 0
                                       	gps_adid         {gps_adid}
                                       	needs_response_details 0
                                       	tracking_enabled 1
12-05 22:59:26.638 5005-8275 V/Adjust: Response: {"timestamp":"2016-12-05T15:59:26.985Z+0000","message":"Attribution found","attribution":{}}
12-05 22:59:26.638 5005-8275 I/Adjust: Attribution found
12-05 22:59:26.642 5005-8275 D/Adjust: Wrote Activity state: ec:0 sc:1 ssc:1 sl:0.0 ts:0.0 la:11:12:13 uuid:{uuid}
12-05 22:59:26.646 5005-8080 D/Adjust: Wrote Attribution: tt:null tn:null net:null cam:null adg:null cre:null cl:null
12-05 22:59:27.457 5005-8080 V/Adjust: Foreground timer suspended with 54.6 seconds left
12-05 22:59:27.457 5005-8080 V/Adjust: Subsession end
12-05 22:59:27.458 5005-8080 D/Adjust: Wrote Activity state: ec:0 sc:1 ssc:1 sl:5.4 ts:5.4 la:11:12:13 uuid:{uuid}
12-05 22:59:29.075 5005-8080 D/Adjust: Wrote Activity state: ec:1 sc:1 ssc:1 sl:5.7 ts:5.7 la:11:12:13 uuid:{uuid}
12-05 22:59:29.075 5005-8080 V/Adjust: Foreground timer starting
12-05 22:59:29.075 5005-8080 V/Adjust: Subsession start
12-05 22:59:29.081 5005-8080 V/Adjust: Started subsession 2 of session 1
12-05 22:59:29.100 5005-8080 D/Adjust: Wrote Activity state: ec:1 sc:1 ssc:2 sl:7.0 ts:5.7 la:11:12:13 uuid:{uuid}
12-05 22:59:29.106 5005-8535 D/Adjust: Added package 1 (event'{event}')
12-05 22:59:29.108 5005-8535 V/Adjust: Path:      /event
                                       ClientSdk: android4.10.4
                                       Parameters:
                                       	android_uuid     {android_uuid}
                                       	api_level        23
                                       	app_token        {token}
                                       	app_version      1.7.1
                                       	attribution_deeplink 1
                                       	country          US
                                       	cpu_type         armeabi-v7a
                                       	created_at       2016-12-05T22:59:27.801Z+0700
                                       	device_manufacturer LGE
                                       	device_name      Nexus 5
                                       	device_type      phone
                                       	display_height   1776
                                       	display_width    1080
                                       	environment      sandbox
                                       	event_buffering_enabled 0
                                       	event_count      1
                                       	event_token      hhwti1
                                       	fb_id            {fb_id}
                                       	gps_adid         {gps_adid}
                                       	hardware_name    M4B30X
                                       	language         en
                                       	needs_response_details 0
                                       	os_build         M4B30X
                                       	os_name          android
                                       	os_version       6.0.1
                                       	package_name    {package}
                                       	screen_density   high
                                       	screen_format    normal
                                       	screen_size      normal
                                       	session_count    1
                                       	session_length   6
                                       	subsession_count 1
                                       	time_spent       6
                                       	tracking_enabled 1
                                       	vm_isa           arm
12-05 22:59:29.139 5005-8535 D/Adjust: Wrote Package queue: [event'{event}']
12-05 22:59:29.139 5005-8535 D/Adjust: Package handler wrote 1 packages
12-05 22:59:29.140 5005-8535 V/Adjust: Package handler is already sending
12-05 22:59:30.186 5005-8539 V/Adjust: Response: {"app_token":"{token}","adid":"{adid}","timestamp":"2016-12-05T15:59:30.517Z+0000","message":"Event tracked '{event}'","attribution":{"tracker_token":"{tracker_token}","tracker_name":"{tracker_name}","network":"{tracker_name}"}}
12-05 22:59:30.186 5005-8539 I/Adjust: Event tracked '{event}'
12-05 22:59:30.196 5005-8626 D/Adjust: Wrote Package queue: []
12-05 22:59:30.196 5005-8626 D/Adjust: Package handler wrote 0 packages
>12-05 22:59:30.196 5005-8626 V/Adjust: Package handler can send

**Comment by uerceg** [comment]
In the attribution request response, I see this in your logs:
```
12-05 22:59:26.638 5005-8275 V/Adjust: Response: {"timestamp":"2016-12-05T15:59:26.985Z+0000","message":"Attribution found","attribution":{}}
```
Is value of `"attribution"` key in this JSON `{}` and that is what you have get as a response?

**Comment by avinz** [comment]
@uerceg Yes, this is from logs, i dint hide there anything :)

**Comment by avinz** [comment]
@uerceg Hey, any news?)

**Comment by uerceg** [comment]
Hey. Yes, can you check again? :) Some of the SDK versions had the old relict behaviour that attribution was not being asked for unless attribution callback was not implemented in the app (which is the case in your app as far as I see from logs). We made backend to give us this info now either way. Can you retry and lemme know if it worked?

**Comment by avinz** [comment]
@uerceg Woohoo, finally works :) Tested several times, opens deeplink immediately after first launch. Thank you!

**Comment by uerceg** [comment]
Great news, glad it worked!

Cheers.

**Comment by avinz** [comment]
@uerceg Hey there again xD
Faced new issue. I've tested on Nexus 5 and 5x, everything works perfect. I receive deeplinks, then I can forget the device by device ID and receive again.
My colleagues tested on Huawei P9 Lite and OnePlus 2 and nothing works. Though they make the same flow, no deeplink and when I request forget with their ID's replies "Device not found". Tested on same app versions.

**Comment by avinz** [comment]
@uerceg Still facing this issue, cant figure out what can be the problem.

**Comment by uerceg** [comment]
Hi @avinz 

Sorry for the delay. Hm. The whole flow shouldn't be related to the device.

Can I get the same logs from the test flow like you sent them few posts above?

**Comment by avinz** [comment]
@uercegI found the issue, should use adid to forget, not the device ID.
So if I understood correct, one device can get deferred deeplink only once without using forget device? Cause I saw in logs on other devices - "Session request check interval failed (Ignoring too frequent session. Last session: 2016-12-13T15:23:54, this session: 2016-12-13T15:30:37, interval: 6m43s, min interval: 20m) (app_token: {token}, {adid}))". But previously user opened the app more than one day before.

**Comment by uerceg** [comment]
Well, this message means that you deleted the app from the device and reinstalled it sooner than 20 minutes after deletion, but you `didn't forget the device from our backend`. New app installation initialises the SDK for the first time within that app installation and our SDK by default sends session package first to the backend. But backend doesn't take too frequent session requests into consideration, thus this one gets rejected. If you forgot the device, our backend wouldn't know about it so this session would be counted as an install.

Deferred deep linking happens only on `app installation`. So install needs to be tracked in order to simulate deferred deep linking scenario in your app. In order to do that, app must be deleted from your device, device successfully forgotten from the backend, URL with `deep_link` parameter clicked and app newly installed on your device (in that exact order). Then, deferred deep link will arrive into the app.

**Comment by avinz** [comment]
@uerceg As i understood.
Deeplink should arrive when :
1)User installs and opens app firstly on device after clicking on adjust link with deeplink
2)User deinstall app, and installs it again after 20 minutes with the click on same adjust link

Is that correct?

**Comment by uerceg** [comment]
1) Yes
2) Yes and no, depends. In this case, there's one additional condition that needs to be fulfilled -> new installation from same user needs to happen outside of `reattribution window` time which is set on your adjust dashboard.

More elaboration on case 2: After your user uninstalls the app, at this moment, we're not marking this uninstall on the backend side, so for our backend, your user still exists with the app on his device. Once he/she installs it again (after 20 minutes++), backend will simply treat this as a new session, not an install. Because of this, if this new session happened outside of reattribution window, user will get reattributed and deferred deep link will arrive to the app.

Uninstall tracking feature is coming soon and after this feature is live, I guess that your guess in scenario 2 will be the flow of deferred deep link arrival into the app after uninstallation and reinstallation.

**Comment by avinz** [comment]
@uerceg Ah, in adjust settings of the app there is "INACTIVE USER DEFINITION", it is set to 7 days, so I suppose in our case should pass 7 days, not 20 minutes. Right?

**Comment by uerceg** [comment]
@avinz Exactly.

**Comment by aydinozkan** [comment]
Hello,

I know this is an old issue but it was very informative and I've tried to follow the steps to solve a similar issue I'm facing. Below is the output:

> V/Adjust: Response: {"error":"Session failed (device is opted out) (app_token: {my_app_token})","tracking_state":"opted_out"}
E/Adjust: No message found

What might be the problem, shouldn't we use "forget device" url any more?

Thanks.

**Comment by uerceg** [comment]
Hi @aydinozkan 

This response from backend indicates that device you're testing is forgotten in accordance to GDPR law. So at some point in time, you triggered this method https://github.com/adjust/android_sdk#gdpr-forget-me on it.

Once GDPR forgotten devices are not possible to be forgotten with forget device option in testing console and reused again, they are simply gone forever.

If you wanna retry testing of things on that particular device, the only way is for you to make your device appear as new one to us - uninstall an app completely from it, reset Google Advertising Identifier and reinstall an app again.

Hopefully, it should work. Let us know if it did.

Cheers

**Comment by aydinozkan** [comment]
Hi @uerceg 

Most probably :) I I reset the gps_adid and its working now. 

Thank you.
